R/binary_classification.R

```diff
@@ -138,3 +140,244 @@ fbeta_score <- function(actual, predicted, beta = 1) {
     rec <- recall(actual, predicted)
     return((1 + beta^2) * prec * rec / ((beta^2 * prec) + rec))
 }
+
+#' Binary confusion matrix
+#'
+#' \code{confusion_matrix} Calculates a binary classification confusion matrix,
+#' comparing the predicted with the actual values for the classes.
+#' Assumes that 1 is used for the positive class and 0 for the
+#' negative class.
+#'
+#' Returns a \code{data.frame} with columns corresponding to the
+#' number of True Positives (\code{tp}), False Positives (\code{fp}),
+#' True Negatives (\code{tn}), and False Negatives (\code{fn})
+#'
+#' @inheritParams params_binary
+#' @seealso \code{\link{sensitivity}} \code{\link{specificity}}
+#' @examples
+#' actual <- c(1, 1, 1, 0, 0, 0)
+#' predicted <- c(1, 0, 1, 1, 1, 1)
+#' confusion_matrix(actual, predicted)
+confusion_matrix <- function(actual, predicted) {
+    binvals <- c(0, 1)
+    b_actual <- unique(actual)
+    b_predicted <- unique(predicted)
+
+    # ideally "actual" should be a combination of 0s and 1s,
+    # but could be all 0s or all 1s as degenerate cases
+    if (!(
+        setequal(binvals, b_actual) |
+        setequal(c(0), b_actual) |
+        setequal(c(1), b_actual)
+    )) {
+        stop(paste("Expecting a vector of 0s and 1s for 'actual'. Got:",
+                   paste(actual, collapse = ", ")))
+    }
+
+    # "predicted" could be all 0s, all 1s, or a combination
+    if (!(
+        setequal(binvals, b_predicted) |
+        setequal(c(0), b_predicted) |
+        setequal(c(1), b_predicted)
+    )) {
+        stop(paste("Expecting a vector of 0s and 1s for 'predicted'. Got:",
+                    paste(predicted, collapse = ", ")))
+    }
+
+    if (length(actual) != length(predicted)) {
+        stop(
+            paste(
+                "Size of 'actual' and 'predicted' are not the same: length(actual):",
+                length(actual), "!= length(predicted):", length(predicted)
+            )
+        )
+    }
+
+    # explicit comparisons
+    data.frame(
+        "tp" = sum(actual == 1 & predicted == 1),
+        "fn" = sum(actual == 1 & predicted == 0),
+        "fp" = sum(actual == 0 & predicted == 1),
+        "tn" = sum(actual == 0 & predicted == 0)
+        )
+}
+
+#' Sensitivity
+#'
+#' \code{sensitivity} calculates the proportion of actual positives (\code{actual} equals 1)
+#' that are correctly identified as such. It is also known as
+#' \code{true positive rate} or \code{recall}.
+#'
+#' @inheritParams params_binary
+#' @seealso \code{\link{confusion_matrix}} \code{\link{specificity}}
+#' @export
+#' @examples
+#' actual <- c(1, 1, 1, 0, 0, 0)
+#' predicted <- c(1, 0, 1, 1, 1, 1)
+#' sensitivity(actual, predicted)
+sensitivity <- function(actual, predicted) {
+    recall(actual, predicted)
```

Comment:
Thanks for implementing sensitivity as a function of recall.

I'm thinking of a better way to implement this so that sensitivity and recall actually are defined as the same function, not just one function that calls another. I want the functions to share documentation too. I'll look into the best ways for doing this.

I'm thinking about

```
foo <- bar <- baz <- function(actual, predicted {
    ...
}
```

But I need to do research into how that impacts R packages and documentation.